src/foam/u2/view/ChoiceView.js

```diff
@@ -281,6 +281,14 @@ foam.CLASS({
         this.dao.select().then(function(s) {
           this.choices = s.array.map(this.objToChoice);
           if ( this.data == null && this.index === -1 ) this.index = this.placeholder ? -1 : 0;
+
+          // When the DAO changes, the data we had selected before might not be
+          // relevant anymore. In that case, we need to find out if the
+          // previously selected data is still in the new choices list. If it
+          // is, keep the selection the same, otherwise reset the state so that
+          // there's no selection since the previous selection doesn't exist in
+          // the new DAO.
+          this.choice = this.findChoiceByData(this.data);
```

Comment:
I don't think you want to do this. There's already a method to handle this that's called: `onChoicesUpdate` which should trigger from line 282. Why do you need this?

Reply:
@mcarcaso But it seems that onChoicesUpdate will not be triggered when we select the first choiceView again.  @TW80000

Reply:
`this.onDetach(this.choices$.sub(this.onChoicesUpdate));`
This means that `onChoicesUpdate` will fire when `choices` changes.
If you look at the code for `onChoicesUpdate`, you'll see that it only updates `this.choice` if `this.choices.length` is not zero. It might be worth considering changing that behaviour from:
```
        if ( this.choices.length ) {
          this.choice = ( d != null && this.findChoiceByData(d) ) || this.defaultValue;
        }
```
to
```
this.choice = this.choices.length && d != null &&
    this.findChoiceByData(d) ) ||
    this.defaultValue;
```

Reply:
Carl and I looked a this offline and there are some subtle issues around this view. The main issue was that the index wasn't getting updated to -1 when the choice was getting cleared so when the user changes the value to the first choice in the list, nothing was happening because index was already 0.